src/imap.rs

```diff
@@ -1179,101 +1181,85 @@ impl Imap {
-                                folder.as_ref(),
-                                uid,
-                                dest_folder.as_ref(),
-                                err
-                            );
-
-                            false
-                        }
-                    }
-                } else {
-                    unreachable!();
-                };
-
-                if !moved {
-                    let copied = if let Some(ref mut session) = &mut *self.session.lock().unwrap() {
-                        match session.uid_copy(&set, &dest_folder) {
-                            Ok(_) => true,
-                            Err(err) => {
-                                eprintln!("error copy: {:?}", err);
-                                info!(context, 0, "Cannot copy message.",);
-
-                                false
-                            }
-                        }
-                    } else {
-                        unreachable!();
-                    };
-
-                    if copied {
-                        if self.add_flag(context, uid, "\\Deleted") == 0 {
-                            warn!(context, 0, "Cannot mark message as \"Deleted\".",);
-                        }
-                        self.config.write().unwrap().selected_folder_needs_expunge = true;
-                        res = DC_SUCCESS;
-                    }
+                ImapResult::RetryLater
+            };
+        }
+        let set = format!("{}", uid);
+        if let Some(ref mut session) = &mut *self.session.lock().unwrap() {
+            match session.uid_mv(&set, &dest_folder) {
+                Ok(_) => {
+                    // XXX set dest_uid properly (like it was done in C)
+                    return ImapResult::Success;
+                }
+                Err(err) => {
+                    info!(
+                        context,
+                        0,
+                        "Cannot move message, fallback to COPY/DELETE {}/{} to {}: {}",
+                        folder.as_ref(),
+                        uid,
+                        dest_folder.as_ref(),
+                        err
+                    );
                 }
             }
-        }
-
-        if res == DC_SUCCESS {
-            // TODO: is this correct?
-            *dest_uid = uid;
-        }
+        } else {
+            unreachable!();
+        };
 
-        if res == DC_RETRY_LATER {
-            if self.should_reconnect() {
-                DC_RETRY_LATER
-            } else {
-                DC_FAILED
+        // message was NOT moved, let's try copy
+        if let Some(ref mut session) = &mut *self.session.lock().unwrap() {
+            match session.uid_copy(&set, &dest_folder) {
+                Ok(_) => {}
+                Err(err) => {
+                    info!(context, 0, "Cannot copy message. {:?}", err);
+                    return ImapResult::Failed;
```

Comment:
maybe we should do a  
```
return if !self.should_reconnect() {
                ImapResult::Failed
            } else {
                ImapResult::RetryLater
            };
```
here as well. otherwise, if network does away during copy, the message is never tried to copy again.

maybe it is also better to add the same check after mv-failing already - if it turnes out, network is away after move, trying over with copy+delete mayl raise the risk in a flaky network situation that one of these part fails.